### PR TITLE
Adding support for hive partitioned external tables in BigQuery

### PR DESCRIPTION
**Pull request by msen92**
Hi,

We are trying to write our application logs to Google Cloud Storage and be able query the data from external tables at BigQuery. We also want to be able to use hive partitioning to reduce the amount of data processed by queries by forcing the use of partition filters when referencing external tables. To support this I added a configuration named `hive_partition_files` and when this configuration is set to true, uploads will parse log files date pattern and upload the files to `log_file_prefix/dt=%Y-%m-%d` directories at the bucket. First I tried to accomplish this by adding forward slashes to `date_pattern` and `log_file_prefix` configuration but it was not possible like mentioned in #47. Also #37 was a similar feature request and I think this PR might help close those issues. I also shared a screen shot of the folder structure and output config that I used to create the log files at gcs for test purposes. As a first time contributor I welcome any feedbacks and ideas.

```
output {
   google_cloud_storage {
     bucket => "bucket"
     json_key_file => "keyfile.json"
     log_file_prefix => "msen_prefix"
     max_file_size_kbytes => 64
     output_format => "json"
     date_pattern => "%Y-%m-%dT%H:00"
     flush_interval_secs => 2
     gzip => true
     gzip_content_encoding => false
     uploader_interval_secs => 1
     hive_partition_files => true
   }
}
``` 

resim

### CODE REVIEW

File: lib/logstash/outputs/google_cloud_storage.rb [hunk 1]
```diff
@@ -154,6 +155,12 @@ class LogStash::Outputs::GoogleCloudStorage < LogStash::Outputs::Base
 
   config :max_concurrent_uploads, :validate  => :number, :default => 5
 
+  # When true, uploaded files will be placed into sub-folders based on their creation dates.
+  # log_file_prefix configuration will be used as the parent folder and
+  # uploaded files will be placed on sub-folders in dt=%Y-%m-%d format.
+  # Foldered files can be used for creating hive partitioned external tables at BigQuery.
+  config :hive_partition_files, :validate  => :boolean, :default => false
```
Comment (mashhurs) [inline lib/logstash/outputs/google_cloud_storage.rb:162]: Instead boolean value (which is very hard to extend in the future), we may directly use the format.
Something like `hive_partition_format`? defaults to no-value, which keeps current behavior.

**Review by mashhurs — CHANGES_REQUESTED** [review]
Beside the comments, changelog.md and version updates required.

File: logstash-output-google_cloud_storage.gemspec [hunk 2]
```diff
@@ -30,6 +30,7 @@ Gem::Specification.new do |s|
   s.add_runtime_dependency 'logstash-codec-json_lines'
   s.add_runtime_dependency 'mime-types', '~> 2' # last version compatible with ruby 2.x
   s.add_runtime_dependency 'concurrent-ruby' # use version bundled with Logstash to avoid platform mismatch on plugin install
+  s.add_runtime_dependency 'thwait', '~> 0.2.0'
```
[the remark this answers is not in the document]
Comment (msen92) [reply logstash-output-google_cloud_storage.gemspec]: Removed it :+1:

File: lib/logstash/outputs/google_cloud_storage.rb (hunk 1 above)
Comment (msen92) [reply lib/logstash/outputs/google_cloud_storage.rb]: Makes so much sense. My approach was completely static. In this way we can supply different partitioning structures easily. I changed the configuration name and partitioning logic as you suggested.

**Pull request closed without merging**